taskcat/s3_sync.py

```diff
@@ -0,0 +1,157 @@
+# Jay McConnell <jmmccon@amazon.com>,
+# Andrew Glenn <andglenn@amazon.com>
+from __future__ import print_function
+
+import hashlib
+import fnmatch
+import os
+import time
+from functools import partial
+from multiprocessing.dummy import Pool as ThreadPool
+from taskcat.colored_console import PrintMsg
+
+
+class S3Sync(object):
+    """Syncronizes local project files with S3 based on checksums.
+
+    Excludes hidden files, unpackaged lambda source and taskcat /ci/ files.
+    Uses the Etag as an md5 which introduces the following limitations
+        * Uses undocumented etag algorithm for multipart uploads
+        * Does not work wil files uploaded in the console that use SSE encryption
+        * see https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html for more info
+    Does not support buckets with versioning enabled
+    """
+
+    _exclude_files = [
+        ".*",
+        "*.md"
+    ]
+    _exclude_path_prefixes = [
+        "functions/source/",
+        ".",
+        "ci/"
+    ]
+
+    def __init__(self, s3_client, bucket, prefix, path, acl=None):
+        """Syncronizes local file system with an s3 bucket/prefix
+
+        """
+        if prefix != "" and not prefix.endswith('/'):
+            prefix = prefix + '/'
+        self.s3_client = s3_client
+        fl = self._get_local_file_list(path)
+        s3fl = self._get_s3_file_list(bucket, prefix)
+        self._sync(fl, s3fl, bucket, prefix, acl=acl)
+        return
+
+    def _hash_file(self, file_path, chunk_size=8 * 1024 * 1024):
+        # This is a bit funky because of the way multipart upload etags are done, they are a md5 of the md5's from each part with the number of parts appended
+        # credit to hyperknot https://github.com/aws/aws-cli/issues/2585#issue-226758933
+        md5s = []
+
+        with open(file_path, 'rb') as fp:
+            while True:
+                data = fp.read(chunk_size)
+                if not data:
+                    break
+                md5s.append(hashlib.md5(data))
+
+        if len(md5s) == 1:
+            return '"{}"'.format(md5s[0].hexdigest())
+
+        digests = b''.join(m.digest() for m in md5s)
+        digests_md5 = hashlib.md5(digests)
+        return '"{}-{}"'.format(digests_md5.hexdigest(), len(md5s))
+
+    def _get_local_file_list(self, path, include_checksums=True):
+        file_list = []
+        path = os.path.abspath(os.path.expanduser(path))
+        for root, dirs, files in os.walk(path):
+            relpath = os.path.relpath(root, path) + "/"
+            exclude_path = False
+            if relpath == './':
+                relpath = ""
+            for p in S3Sync._exclude_path_prefixes:
+                if relpath.startswith(p):
+                    exclude_path = True
+                    break
+            if not exclude_path:
+                for file in files:
+                    exclude = False
+                    for p in S3Sync._exclude_files:
```

Comment:
Can this be refactored? Looping over the `S3Sync._exclude_files` list for each file seems costly.

Reply:
done